grpc_adapter.go

```diff
@@ -0,0 +1,668 @@
+
+	w.seenHeaders = keys
+	w.wroteHeader = true
+}
+
+func (w *grpcWebResponseWriter) Write(b []byte) (int, error) {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	return w.ResponseWriter.Write(b)
+}
+
+func (w *grpcWebResponseWriter) WriteHeader(code int) {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	w.ResponseWriter.WriteHeader(code)
+}
+
+func (w *grpcWebResponseWriter) flushWithTrailers() {
+	if !w.wroteHeader {
+		w.writeHeaders()
+	}
+	if err := w.writeTrailer(); err != nil {
+		return // nothing
+	}
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *grpcWebResponseWriter) Flush() {
+	flushResponseWriter(w.ResponseWriter)
+}
+
+func (w *grpcWebResponseWriter) writeTrailer() error {
+	header := w.Header()
+	trailer := make(http.Header, len(header)-len(w.seenHeaders))
+	for key, values := range header {
+		if w.seenHeaders[key] {
+			continue
+		}
+		key = strings.TrimPrefix(key, http.TrailerPrefix)
+		// https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2
+		trailer[strings.ToLower(key)] = values
+	}
+
+	var buf bytes.Buffer
+	if err := trailer.Write(&buf); err != nil {
+		return err
+	}
+
+	head := [5]byte{1 << 7, 0, 0, 0, 0} // MSB=1 indicates this is a trailer data frame.
+	binary.BigEndian.PutUint32(head[1:5], uint32(buf.Len()))
+	if _, err := w.ResponseWriter.Write(head[:]); err != nil {
+		return err
+	}
+	if _, err := w.ResponseWriter.Write(buf.Bytes()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// bufferedEnvelopeReader buffers a message to wrap it in an envelope.
+type bufferedEnvelopeReader struct {
+	io.ReadCloser
+
+	config     *grpcAdapterConfig
+	buf        bytes.Buffer
+	buffered   bool
+	compressed bool
+	err        error
+}
+
+func (r *bufferedEnvelopeReader) fillBuffer() error {
+	body, err := readAll(nil, r.ReadCloser, int64(r.config.ReadMaxBytes))
+	if err != nil {
+		// Limit reached if EOF.
+		if errors.Is(err, io.EOF) {
+			bytesRead := int64(len(body))
+			discardedBytes, err := io.Copy(io.Discard, r.ReadCloser)
+			if err != nil {
+				return errorf(CodeResourceExhausted,
+					"message is larger than configured max %d - unable to determine message size: %w",
+					r.config.ReadMaxBytes, err)
+			}
+			return errorf(CodeResourceExhausted,
+				"message size %d is larger than configured max %d",
+				bytesRead+discardedBytes, r.config.ReadMaxBytes)
+		}
+		return errorf(CodeUnknown, "read message: %w", err)
+	}
+	size := uint32(len(body))
+
+	var head [5]byte
+	if r.compressed {
+		head[0] = 1
+	}
+	binary.BigEndian.PutUint32(head[1:], size)
+	r.buf.Write(head[:])
+	r.buf.Write(body)
+	return nil
+}
+
+func (r *bufferedEnvelopeReader) Read(data []byte) (int, error) {
+	if !r.buffered {
+		r.err = r.fillBuffer()
+		r.buffered = true
+	}
+	if r.err != nil {
+		return 0, r.err
+	}
+	return r.buf.Read(data)
+}
+
+// readAll reads from r until an error or the limit is reached.
+func readAll(data []byte, reader io.Reader, limit int64) ([]byte, error) {
```

Comment:
I think this could be simplified and leverage the std lib a bit more
1. If there is no limit, simply return `io.ReadAll`
2. If there is a limit, create an `io.LimitReader` with the limit == `limit`. Then call `io.ReadAll`. If the resulting data length == limit, call `reader.Read` to get one byte, and if it's not EOF, then the incoming data is too large.